vim: Add some forced motion support

mastion opened the pull request:
Closes https://github.com/zed-industries/zed/issues/20971

Added `v` input to yank and delete to override default motion. The global vim state tracking if the forced motion flag was passed handled the same way that the count is. [The main chunk of code maps the motion kind from the default to the overridden kind](https://github.com/zed-industries/zed/pull/27991/files#diff-2dca6b7d1673c912d14e4edc74e415abbe3a4e6d6b37e0e2006d30828bf4bb9cR1249-R1254). To handle the case of deleting a single character (dv0) at the start of a row I had to modify the control flow [here](https://github.com/zed-industries/zed/pull/27991/files#diff-2dca6b7d1673c912d14e4edc74e415abbe3a4e6d6b37e0e2006d30828bf4bb9cR1240-R1244). Then to handle an exclusive delete till the end of the row (dv$) I [saturated the endpoint with a left bias](https://github.com/zed-industries/zed/pull/27991/files#diff-2dca6b7d1673c912d14e4edc74e415abbe3a4e6d6b37e0e2006d30828bf4bb9cR1281-R1286).

Test case: dv0

https://github.com/user-attachments/assets/613cf9fb-9732-425c-9179-025f3e107584

Test case: yvjp

https://github.com/user-attachments/assets/550b7c77-1eb8-41c3-894b-117eb50b7a5d

Release Notes:

- Added some forced motion support for delete and yank

mastion committed:
WIP: first pass. tests work. code is messy

mastion committed:
only set the motionkind when inclusive is true and motionkind is none

mastion committed:
removes delete2 and use operation::delete

mastion committed:
fix operation display

was showing as ddv now shows as dv

mastion committed:
changed test name and needed to change the neovim cached data

mastion commented:
A note about implementation -
My first thought when working on this was to set the `MotionKind` to `Inclusive` up at the `PushDeleteInclusive` function ( or somewhere near that ) but as I dug through the code I noticed inside of the [range](https://github.com/zed-industries/zed/blob/main/crates/vim/src/motion.rs#L1143) function in `motion.rs` is defaults to `Exclusive`. 
I didn't feel like I had the context for changing that, so I just passed in a bool.

KyleBarton commented on crates/vim/src/state.rs:
```diff
@@ -78,7 +78,11 @@ impl Default for Mode {
 pub enum Operator {
     Change,
-    Delete,
+    //Delete,
```
Should delete these comments

KyleBarton commented on crates/vim/src/normal/delete.rs:
```diff
@@ -33,12 +34,20 @@ impl Vim {
                         ranges_to_copy
                             .push(selection.start.to_point(map)..selection.end.to_point(map));
 
+                        if inclusive {
```
any reason this needs to go below `ranges_to_copy`? If not, I think you could avoid `mut` with:

```
let kind = match motion.expand_selection(...) {
  Some(x) => Some(x)
  None => {
     if (inclusive) {
       {selection mutating code}
       Some(MotionKind::Inclusive)
     } else {
        None
     }
  }
}
```

But no strong opinion on that, depending on if you think that reads worse.

mastion replied:
To do it that way the code would look closer to 
```
let kind = match motion.expand_selection(...) {
  Some(x) if inclusive => {
    {selection mutation code}
    Some(x)
  },
  None if inclusive => {
       {selection mutating code}
       Some(MotionKind::Inclusive)
  },
  Some(x) => Some(x),
  None => None
}
```
I split the difference and was able to keep the kind from being a `mut` but the mutation code happens after the fact.

KyleBarton commented on crates/vim/src/state.rs:
```diff
@@ -933,7 +937,8 @@ impl Operator {
             Operator::Change => "c",
-            Operator::Delete => "d",
+            Operator::Delete { inclusive: false } => "d",
+            Operator::Delete { inclusive: true } => "v",
```
I'm lower context on this - it seems fine? My only question would be, could future edits to `vim.json` spuriously allow a user to directly access the `PushDeleteInclusive` operator, rather than traveling first through the `PushDelete` operator? And if so, is that bad?

mastion replied:
If i'm understanding the question correctly then yeah I think so. 
Someone would just need to move the [pushdeleteinclusive mapping](https://github.com/zed-industries/zed/pull/27991/files/8cc430f023f18a2e68e6cc8404faee8fc72cf8b4#diff-7b7dabbece7a54e041c4b6ab8f69afbe994d301dbded14921a31598d7f6de287R456) out from under the [d operator context](https://github.com/zed-industries/zed/pull/27991/files/8cc430f023f18a2e68e6cc8404faee8fc72cf8b4#diff-7b7dabbece7a54e041c4b6ab8f69afbe994d301dbded14921a31598d7f6de287L452).
But that is true for anything in that section so I don't think it is concern.

mastion committed:
cleanup comment

mastion committed:
refactor: removes mu from kind, do inclusive logic in match

ConradIrwin commented:
@mastion Thanks for this! 

The `v` affects any motion, not just delete; so I think we should treat `v` (and `shift-v`) in the same way we treat the count. `PushInclusive` would set a field on the vim state, which we'd take and pass down into `motion.expand_selection` and `motion.range`  alongside `count`. We'd use that to modify the kind returned from `default_kind()`, and (hopefully) hit our existing inclusive/exclusive code handling (which was tidied up a bunch in #27786).

In particular I would like to make `y v j` (linewise copy becomes exclusive) and `d v e` (inclusive delete becomes exclusive) work in addition to `d v 0` as part of this change.

That said, I don't know if the theory actually works in practice, so happy to pair on this if you want: https://cal.com/conradirwin/pairing

```
v               When used after an operator, before the motion command: Force
                the operator to work charwise, also when the motion is
                linewise.  If the motion was linewise, it will become
                exclusive.
                If the motion already was charwise, toggle
                inclusive/exclusive.  This can be used to make an exclusive
                motion inclusive and an inclusive motion exclusive.
```

mastion commented:
@ConradIrwin Awesome. That is what I expected, but wasn't sure how to structure it.
I'll take a crack at it and schedule some time with you next week in case it doesn't work out as hoped.

mastion committed:
delete tests passing. need to build out rest of inclusive overide logic

mastion committed:
WIP: Added tests for `d v e` and `y v j`

also renamed some variables to match the vim actions

mastion committed:
turn forced_motion off and test it

mastion committed:
testing that forced_motion is off in the other forced motion tests

mastion committed:
change test name and rename cached neovim data

mastion committed:
Paired down test cases, rebased for bug fix, forced_motion following count pattern

Paired with @ConradIrwin to get the tests working/removing nonsense tests cases

mastion committed:
move start of line logic out of move_point and into range

mastion committed:
removed force_motion param from substitute flow as it doesnt take in a motion

also added test case for `d v $` delete + forced motion + to end of line and made it work

ConradIrwin commented:
Thanks for this!
